Mathlib/LinearAlgebra/Prod.lean

```diff
@@ -883,137 +882,22 @@ end LinearMap
-
-theorem tailing_le_tunnel (f : M × N →ₗ[R] M) (i : Injective f) (n : ℕ) :
-    tailing f i n ≤ OrderDual.ofDual (α := Submodule R M) (tunnel f i n) := by
-  dsimp [tailing, tunnelAux]
-  rw [Submodule.map_comp, Submodule.map_comp]
-  apply Submodule.map_subtype_le
-#align linear_map.tailing_le_tunnel LinearMap.tailing_le_tunnel
-
-theorem tailing_disjoint_tunnel_succ (f : M × N →ₗ[R] M) (i : Injective f) (n : ℕ) :
-    Disjoint (tailing f i n) (OrderDual.ofDual (α := Submodule R M) <| tunnel f i (n + 1)) := by
-  rw [disjoint_iff]
-  dsimp [tailing, tunnel, tunnel']
-  erw [Submodule.map_inf_eq_map_inf_comap,
-    Submodule.comap_map_eq_of_injective (tunnelAux_injective _ i _), inf_comm,
-    Submodule.fst_inf_snd, Submodule.map_bot]
-#align linear_map.tailing_disjoint_tunnel_succ LinearMap.tailing_disjoint_tunnel_succ
-
-theorem tailing_sup_tunnel_succ_le_tunnel (f : M × N →ₗ[R] M) (i : Injective f) (n : ℕ) :
-    tailing f i n ⊔ (OrderDual.ofDual (α := Submodule R M) $ tunnel f i (n + 1)) ≤
-      (OrderDual.ofDual (α := Submodule R M) <| tunnel f i n) := by
-  dsimp [tailing, tunnel, tunnel', tunnelAux]
-  erw [← Submodule.map_sup, sup_comm, Submodule.fst_sup_snd, Submodule.map_comp, Submodule.map_comp]
-  apply Submodule.map_subtype_le
-#align linear_map.tailing_sup_tunnel_succ_le_tunnel LinearMap.tailing_sup_tunnel_succ_le_tunnel
-
-/-- The supremum of all the copies of `N` found inside the tunnel. -/
-def tailings (f : M × N →ₗ[R] M) (i : Injective f) : ℕ → Submodule R M :=
-  partialSups (tailing f i)
-#align linear_map.tailings LinearMap.tailings
-
-@[simp]
-theorem tailings_zero (f : M × N →ₗ[R] M) (i : Injective f) : tailings f i 0 = tailing f i 0 := by
-  simp [tailings]
-#align linear_map.tailings_zero LinearMap.tailings_zero
-
-@[simp]
-theorem tailings_succ (f : M × N →ₗ[R] M) (i : Injective f) (n : ℕ) :
-    tailings f i (n + 1) = tailings f i n ⊔ tailing f i (n + 1) := by simp [tailings]
-#align linear_map.tailings_succ LinearMap.tailings_succ
-
-theorem tailings_disjoint_tunnel (f : M × N →ₗ[R] M) (i : Injective f) (n : ℕ) :
-    Disjoint (tailings f i n) (OrderDual.ofDual (α := Submodule R M) <| tunnel f i (n + 1)) := by
-  induction' n with n ih
-  · simp only [tailings_zero]
-    apply tailing_disjoint_tunnel_succ
-  · simp only [tailings_succ]
-    refine Disjoint.disjoint_sup_left_of_disjoint_sup_right ?_ ?_
-    · apply tailing_disjoint_tunnel_succ
-    · apply Disjoint.mono_right _ ih
-      apply tailing_sup_tunnel_succ_le_tunnel
-#align linear_map.tailings_disjoint_tunnel LinearMap.tailings_disjoint_tunnel
-
-theorem tailings_disjoint_tailing (f : M × N →ₗ[R] M) (i : Injective f) (n : ℕ) :
-    Disjoint (tailings f i n) (tailing f i (n + 1)) :=
-  Disjoint.mono_right (tailing_le_tunnel f i _) (tailings_disjoint_tunnel f i _)
-#align linear_map.tailings_disjoint_tailing LinearMap.tailings_disjoint_tailing
-
-end Tunnel
+-- NOTE: the tunnel and tailing APIs are removed
```

Comment:
Could you please add some instructions/pointers about how they are superseded?
Would it make sense to leave the API in place for now, but add `@[deprecated "some note"]` to these declarations for a while?

Reply:
> Could you please add some instructions/pointers about how they are superseded?

That's the problem. They are not superseded, but only removed. There are no direct replacements for them. In the whole mathlib, they were only used in the proof of `IsNoetherian.equivPUnitOfProdInjective`. But now it's proved by `IsNoetherian.injective_of_surjective_of_injective` which uses `iterateMapComap`. This makes them not used by mathlib anymore. There are some discussions around <https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2312076.20Orzech's.20theorem/near/439460160> and <https://github.com/leanprover-community/mathlib4/pull/12076#discussion_r1597680813>.

> Would it make sense to leave the API in place for now

I think maybe it's better. Just leave these for now; remove them or not is a future PR.